### DISCUSSION TITLE
[Question] applied_torque vs root_physx_view.get_measured_joint_efforts

### DISCUSSION THREAD

**Discussion by Jdowdy05** [isaac-sim/IsaacLab#1867] [Q&A]
### applied_torque vs get_measured_joint_efforts vs get_measured_joint_forces

Hello, I am checking my understanding of the following three functions and what they are returning. 

"Applied_torque" is the calculated torque from the actuator that does not account for any external wrenches on the actuator.

"[get_measured_joint_efforts](https://docs.omniverse.nvidia.com/py/isaacsim/source/extensions/omni.isaac.core/docs/index.html?highlight=get_measured_#omni.isaac.core.articulations.Articulation.get_measured_joint_efforts)" is the external torque applied to the joints from the environment. It is essentially the "[get_measured_joint_forces](https://docs.omniverse.nvidia.com/py/isaacsim/source/extensions/omni.isaac.core/docs/index.html?highlight=get_measured_#omni.isaac.core.articulations.Articulation.get_measured_joint_forces)" wrench converted to joint space for the articulation. Im also assuming that the "get_measured_joint_efforts" is just the reaction torque and not a sum of applied_torque

I have read the documentation for both Isaac lab and Isaac sim a few times, but I am still a bit confused about "get_measured_joint_efforts," as it doesn't explicitly say you receive the reaction forces in torque like "get_measured_joint_forces"  does.

Another question along the same line (though I think I know the answer): The observation function "body_incoming_wrench" has a function call inside called "get_link_incoming_joint_force". Does this give a future reaction force by some +sim.dt ahead?

Thank you in advance for the help!

**Comment by RandomOakForest** [comment]
Thank you for posting this. I'll review with the team and post here any findings.

**Comment by heleiduan** [comment]
Referring to this [API doc](https://docs.omniverse.nvidia.com/py/isaacsim/source/extensions/omni.isaac.core/docs/index.html?highlight=get_measured_joint_efforts#omni.isaac.core.articulations.ArticulationView.get_measured_joint_efforts), it seems that `get_measured_joint_efforts()` and `get_max_efforts()` from `ArticulatioView` cannot be accessed properly with the following errors, 
```  
torques = asset.data._root_physx_view.get_measured_joint_efforts()[:, asset_cfg.joint_ids]
AttributeError: 'ArticulationView' object has no attribute 'get_measured_joint_efforts'
```
While others such as `get_mass_matrices()` works properly.

**Reply by BowenFF** [reply]
are there some solutions?

**Comment by RandomOakForest** [comment]
Thank you for following up. I will move this to our _Discussions_ section for the team to address.

**Comment by RandomOakForest** [comment]
Following up on this, the differences between these function is as follows:

[get_applied_joint_efforts](https://docs.isaacsim.omniverse.nvidia.com/4.5.0/py/source/extensions/isaacsim.core.api/docs/index.html#isaacsim.core.prims.ArticulationView.get_applied_joint_efforts) API will return a tensor that specifies the efforts set by the user through the [set_joint_efforts](https://docs.isaacsim.omniverse.nvidia.com/4.5.0/py/source/extensions/isaacsim.core.api/docs/index.html#isaacsim.core.prims.ArticulationView.set_joint_efforts).

[get_measured_joint_forces](https://docs.isaacsim.omniverse.nvidia.com/4.5.0/py/source/extensions/isaacsim.core.api/docs/index.html#isaacsim.core.prims.ArticulationView.get_measured_joint_forces) API will return a tensor that specifies 6-dimensional spatial forces per joints for all articulations (total overall joint forces). To mimic force-torque sensors, this API can be used to retrieve forces from a fixed joint.

[get_measured_joint_efforts](https://docs.isaacsim.omniverse.nvidia.com/4.5.0/py/source/extensions/isaacsim.core.api/docs/index.html#isaacsim.core.prims.ArticulationView.get_measured_joint_efforts) API will return a tensor which specifies the active components (the projection of the joint forces on the motion direction) of the joint forces for all the joints and articulations.

See [this doc for reference](https://docs.isaacsim.omniverse.nvidia.com/4.5.0/sensors/isaacsim_sensors_physics_articulation_force.html). 

Let us know if you need further help.

**Reply by Zhihaibi** [reply]
Thanks for the answer, but when I use get_measured_joint_efforts(), it always returns [0, 0, 0, 0, 0, 0] when the manipulator is static. But it shouldn't be zero — there should still be torques required to counteract gravity. Is it because this function doesn't include the torques needed for gravity compensation?
